crates/in_stream/src/tcp.rs

```diff
@@ -0,0 +1,323 @@
+use crate::*;
+use net2::TcpStreamExt;
+use std::io::{Error, ErrorKind, Read, Result, Write};
+use url2::prelude::*;
+
+const SCHEME: &'static str = "tcp";
+
+/// internal helper convert urls to socket addrs for binding / connection
+fn tcp_url_to_socket_addr(url: &Url2) -> Result<std::net::SocketAddr> {
+    if url.scheme() != SCHEME || url.host_str().is_none() || url.port().is_none() {
+        return Err(Error::new(
+            ErrorKind::InvalidInput,
+            format!("got: '{}', expected: '{}://host:port'", SCHEME, url),
+        ));
+    }
+    let rendered = format!("{}:{}", url.host_str().unwrap(), url.port().unwrap());
+    match rendered.parse() {
+        Ok(addr) => Ok(addr),
+        Err(_) => Err(Error::new(
+            ErrorKind::InvalidInput,
+            format!("could not parse '{}', as 'host:port'", rendered),
+        )),
+    }
+}
+
+#[derive(Debug)]
+/// configuration options for the listener bind call
+pub struct TcpBindConfig {
+    pub backlog: i32,
+}
+
+impl Default for TcpBindConfig {
+    fn default() -> Self {
+        Self { backlog: 1024 }
+    }
+}
+
+impl InStreamConfig for TcpBindConfig {}
+
+/// basic tcp socket server/listener
+#[derive(Debug)]
+pub struct InStreamListenerTcp(pub std::net::TcpListener);
+
+impl InStreamListenerTcp {
+    pub fn bind(url: &Url2, config: TcpBindConfig) -> Result<Self> {
+        InStreamListenerTcp::raw_bind(url, config)
+    }
+}
+
+impl InStreamListener<&mut [u8], &[u8]> for InStreamListenerTcp {
+    type Stream = InStreamTcp;
+
+    fn raw_bind<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self> {
+        let config = TcpBindConfig::from_gen(config)?;
+        let addr = tcp_url_to_socket_addr(url)?;
+        let listener = net2::TcpBuilder::new_v4()?
+            .bind(addr)?
+            .listen(config.backlog)?;
+        listener.set_nonblocking(true)?;
+        Ok(Self(listener))
+    }
+
+    fn binding(&self) -> Url2 {
+        let local = self.0.local_addr().unwrap();
```

Comment:
How sensitive is this unwrap()?

Reply:
hard to know for sure... it looks like it's mainly this [underlying inet family match](https://github.com/rust-lang/rust/blob/master/src/libstd/sys_common/net.rs#L127-L141):

```rust
    match storage.ss_family as c_int {
        c::AF_INET => {
            assert!(len as usize >= mem::size_of::<c::sockaddr_in>());
            Ok(SocketAddr::V4(FromInner::from_inner(unsafe {
                *(storage as *const _ as *const c::sockaddr_in)
            })))
        }
        c::AF_INET6 => {
            assert!(len as usize >= mem::size_of::<c::sockaddr_in6>());
            Ok(SocketAddr::V6(FromInner::from_inner(unsafe {
                *(storage as *const _ as *const c::sockaddr_in6)
            })))
        }
        _ => Err(Error::new(ErrorKind::InvalidInput, "invalid argument")),
    }
```

Reply:
```suggestion
        let local = self.0.local_addr().expect("Couldn't unwrap local_addr() of TcpListener when trying to get binding URL");
```
?